### PR TITLE
[CR-20] Fix history for chat with Multiple choices 

### PR DESCRIPTION
**Pull request by alexbojko**
#### PROBLEM
Described in CR-20 in jira.
After implementing MultChoices feature chat history is gonna be broken for questions where MC was used. 

---

#### SOLUTION
Fix it somehow

### CODE REVIEW

File: mysite/chat/api.py [hunk 1]
```diff
@@ -1,5 +1,7 @@
 from django.http.response import Http404
 import injections
+from itertools import chain
```
Comment (cmltaWt0) [inline mysite/chat/api.py]: Pep8 imports please.

File: mysite/chat/api.py [hunk 2]
```diff
@@ -74,6 +76,16 @@ class MessagesView(ValidateMixin, generics.RetrieveUpdateAPIView, viewsets.Gener
     authentication_classes = (SessionAuthentication,)
     permission_classes = (IsAuthenticated, IsOwner)
 
+    def roll_fsm_forward(self, chat, message):
+        chat.next_point = self.next_handler.next_point(
```
Comment (cmltaWt0) [inline mysite/chat/api.py]: Add docstring to describe why do we need this method.
Thanks.

File: mysite/chat/api.py [hunk 3]
```diff
@@ -90,12 +102,12 @@ def retrieve(self, request, *args, **kwargs):
...
         ):
             chat.next_point = self.next_handler.next_point(
-                current=message.content, chat=chat, message=message, request=request
+                current=message.content, chat=chat, message=message, request=self.request
             )
             chat.save()
             message.chat = chat
-            serializer = self.get_serializer(message)
-            return Response(serializer.data)
+            # should be replaced with next line:
```
Comment (cmltaWt0) [inline mysite/chat/api.py]: When it should be replaced?

File: mysite/chat/api.py [hunk 4]
```diff
@@ -162,6 +174,48 @@ def perform_update(self, serializer):
                 serializer.save(content_id=resp.id, timestamp=timezone.now(), chat=chat)
             else:
                 serializer.save()
+
+        if (
```
Comment (cmltaWt0) [inline mysite/chat/api.py]: OMG. 😮 
Add TODO to refactor this.

File: mysite/chat/api.py [hunk 5]
```diff
@@ -162,6 +174,48 @@ def perform_update(self, serializer):
...
+                except KeyError:
+                    selected_msg_ids = self.request.data.get(
+                        'selected'
+                    ).keys()
+                    msg_ids = Message.objects.filter(id__in=selected_msg_ids, chat=chat).values_list('id', flat=True)
+                    correct_ids = set(msg_ids).intersection(set(int(i) for i in selected_items.keys()))
+                    selected_choices = []
+                    for i in correct_ids:
+                        selected_choices.append(selected_items[str(i)]['choices'])
+                    selected = chain(*selected_choices)
```
Comment (cmltaWt0) [inline mysite/chat/api.py]: Please add an example of a data we playing with.
It can break my mind ))

File: mysite/chat/api.py [hunk 6]
```diff
@@ -162,6 +174,48 @@ def perform_update(self, serializer):
...
+
+                resp_text = '[selected_choices] ' + ' '.join(str(i) for i in selected)
+            # if not message.content_id:
+            resp = StudentResponse(text=resp_text)
```
Comment (cmltaWt0) [inline mysite/chat/api.py]: Holly Krishna :)

File: mysite/chat/models.py [hunk 7]
```diff
@@ -202,6 +203,48 @@ def get_errors(self):
             errors or '<li><h3>There are no Misunderstands to display.</h3></li>'
         )
 
+    def render_choices(self, choices, checked_choices):
+        choices_template = (
```
Comment (cmltaWt0) [inline mysite/chat/models.py]: Add docstring for this.

File: mysite/chat/models.py [hunk 8]
```diff
@@ -202,6 +203,48 @@ def get_errors(self):
...
+            starmap(lambda i, x: choices_template % (
+                'chat-selectable-selected' if i in checked_choices else '',
+                i,
+                x[2:] if x.startswith(Lesson.NOT_CORRECT_CHOICE) else x[3:]),
+                choices
+            )
+        ) + '</ul>'
+
+    def render_my_choices(self):
+        if '[selected_choices]' in self.content.text:
```
Comment (cmltaWt0) [inline mysite/chat/models.py]: Add docstring and describe a difference with previous method.

File: mysite/fsm/mixins.py [hunk 9]
```diff
@@ -247,11 +260,9 @@ def get_message(self, chat, current=None, message=None):
                 userMessage=True,
                 is_additional=is_additional,
             )
-            if not self.fsm.fsm_name_is_one_of('live_chat'):
-                message = Message.objects.get_or_create(**_data)[0]
-            else:
-                message = Message(**_data)
-                message.save()
+            # here was Message.objects.create for all fsm's except love_chat. for live_chat fsm here was get_or_create
```
Comment (cmltaWt0) [inline mysite/fsm/mixins.py]: `love_chat` :)

File: mysite/fsm/mixins.py [hunk 10]
```diff
@@ -297,9 +308,7 @@ def get_message(self, chat, current=None, message=None):
                 userMessage=True,
                 is_additional=is_additional
             )
-            # if not self.fsm.name == 'live_chat':
-            #     message = Message.objects.get_or_create(**_data)[0]
-            # else:
+            # here was Message.objects.create for all fsm's except love_chat. for live_chat fsm here was get_or_create
```
Comment (cmltaWt0) [inline mysite/fsm/mixins.py]: Same as above.

**Review by cmltaWt0 — CHANGES_REQUESTED** [review]
Tests failed.

File: mysite/chat/api.py (hunk 1 above)
Comment (alexbojko) [reply mysite/chat/api.py]: k

File: mysite/fsm/mixins.py (hunk 9 above)
Comment (alexbojko) [reply mysite/fsm/mixins.py]: yeah love chat.... we are working on Love chat :)

**Pull request merged**